File: backend/prisma/seed.ts
import argon2 from "argon2";
import { PlatformRole, Prisma, PrismaClient, UserStatus } from "@prisma/client";

import {
  PERMISSION_CODES,
  type PermissionCode,
  ROLE_PERMISSION_GRANTS,
  ROLE_CODES,
  type RoleCode,
} from "../src/security/authorization-catalog.js";

const database = new PrismaClient();

const roleDescriptions: Record<RoleCode, { displayName: string; description: string }> = {
  ROOT_ADMIN: {
    displayName: "Root Administrator",
    description: "Institution owner with controlled administrative authority.",
  },
  ADMISSION_ADMIN: {
    displayName: "Admission Administrator",
    description: "Manages student, teacher, and staff onboarding without privilege-granting authority.",
  },
  FINANCE_ADMIN: {
    displayName: "Finance Administrator",
    description: "Views and manages institution finance workflows.",
  },
  PRINCIPAL: {
    displayName: "Principal",
    description: "Manages institution academic structure and views institutional records.",
  },
  TEACHER: {
    displayName: "Teacher",
    description: "Employee role for academic teaching responsibilities.",
  },
  STUDENT: {
    displayName: "Student",
    description: "Student membership within an institution.",
  },
  STAFF: {
    displayName: "Staff",
    description: "General non-teaching employee membership.",
  },
};

const permissionDescriptions: Record<PermissionCode, string> = {
  INSTITUTION_VIEW: "View institution information.",
  INSTITUTION_SETTINGS_MANAGE: "Manage institution settings.",
  ADMIN_VIEW: "View institution administrators.",
  ADMIN_CREATE: "Invite an approved institution administrator role.",
  ADMIN_DISABLE: "Suspend an institution administrator membership.",
  ADMIN_ASSIGN_ROLE: "Assign an explicitly approved institution administrator role.",
  STUDENT_CREATE: "Create student onboarding records.",
  STUDENT_VIEW: "View tenant-scoped student records.",
  STUDENT_UPDATE: "Update tenant-scoped student records.",
  TEACHER_CREATE: "Create teacher onboarding records.",
  TEACHER_VIEW: "View tenant-scoped teacher records.",
  TEACHER_UPDATE: "Update tenant-scoped teacher records.",
  STAFF_CREATE: "Create staff onboarding records.",
  STAFF_VIEW: "View tenant-scoped staff records.",
  STAFF_UPDATE: "Update tenant-scoped staff records.",
  ADMISSION_INVITE_SEND: "Send admissions invitations.",
  ADMISSION_INVITE_RESEND: "Rotate and resend pending admissions invitations.",
  ADMISSION_INVITE_REVOKE: "Revoke pending admissions invitations.",
  ACADEMIC_SESSION_MANAGE: "Manage academic sessions.",
  CLASS_SECTION_MANAGE: "Manage class sections.",
  FINANCE_VIEW: "View finance data.",
  FINANCE_MANAGE: "Manage finance workflows.",
  AUDIT_VIEW: "View institution audit history.",
};

function stableUuid(kind: "role" | "permission", index: number): string {
  const family = kind === "role" ? "4000" : "4001";
  return `00000000-0000-${family}-8000-${String(index + 1).padStart(12, "0")}`;
}

async function seedAuthorizationCatalog(client: Prisma.TransactionClient): Promise<void> {
  const permissions = new Map<PermissionCode, string>();
  for (const [index, permissionCode] of PERMISSION_CODES.entries()) {
    const permission = await client.permission.upsert({
      where: { permissionCode },
      update: { description: permissionDescriptions[permissionCode] },
      create: {
        permissionId: stableUuid("permission", index),
        permissionCode,
        description: permissionDescriptions[permissionCode],
      },
    });
    permissions.set(permissionCode, permission.permissionId);
  }

  for (const [index, roleCode] of ROLE_CODES.entries()) {
    const role = await client.role.upsert({
      where: { roleCode },
      update: roleDescriptions[roleCode],
      create: {
        roleId: stableUuid("role", index),
        roleCode,
        ...roleDescriptions[roleCode],
      },
    });

    await client.rolePermission.deleteMany({ where: { roleId: role.roleId } });
    await client.rolePermission.createMany({
      data: ROLE_PERMISSION_GRANTS[roleCode].map((permissionCode) => ({
        roleId: role.roleId,
        permissionId: permissions.get(permissionCode)!,
      })),
      skipDuplicates: true,
    });
  }
}

async function seedBoards(client: Prisma.TransactionClient): Promise<void> {
  const boards = [
    ["CBSE", "Central Board of Secondary Education"],
    ["CISCE", "Council for the Indian School Certificate Examinations"],
  ] as const;
  for (const [boardCode, displayName] of boards) {
    await client.board.upsert({
      where: { boardCode },
      update: { displayName },
      create: { boardCode, displayName },
    });
  }
}

async function seedPlatformAdministrator(client: Prisma.TransactionClient): Promise<void> {
  const email = process.env.SEED_PLATFORM_ADMIN_EMAIL?.trim().toLowerCase();
  const password = process.env.SEED_PLATFORM_ADMIN_PASSWORD;
  if (!email && !password) return;
  if (!email || !password || password.length < 14) {
    throw new Error(
      "SEED_PLATFORM_ADMIN_EMAIL and a password of at least 14 characters must be supplied together",
    );
  }

  const passwordHash = await argon2.hash(password, { type: argon2.argon2id });
  const user = await client.user.upsert({
    where: { email },
    update: { status: UserStatus.ACTIVE },
    create: {
      email,
      passwordHash,
      status: UserStatus.ACTIVE,
      emailVerifiedAt: new Date(),
      profile: {
        create: {
          firstName: process.env.SEED_PLATFORM_ADMIN_FIRST_NAME?.trim() || "Platform",
          lastName: process.env.SEED_PLATFORM_ADMIN_LAST_NAME?.trim() || "Administrator",
        },
      },
    },
  });

  await client.platformUserRole.upsert({
    where: {
      userId_role: {
        userId: user.userId,
        role: PlatformRole.PLATFORM_SUPER_ADMIN,
      },
    },
    update: {},
    create: {
      userId: user.userId,
      role: PlatformRole.PLATFORM_SUPER_ADMIN,
    },
  });
}

async function main(): Promise<void> {
  await database.$transaction(async (client) => {
    await seedAuthorizationCatalog(client);
    await seedBoards(client);
    await seedPlatformAdministrator(client);
  });
}

main()
  .catch((error: unknown) => {
    console.error(error);
    process.exitCode = 1;
  })
  .finally(async () => {
    await database.$disconnect();
  });

File: backend/scripts/seed.py
from __future__ import annotations
import os
from uuid import uuid4
from sqlalchemy import text
from app.authorization import PERMISSION_DETAILS, ROLE_DETAILS, ROLE_GRANTS
from app.database import transaction
from app.security import hash_password


def stable_uuid(family: str, index: int) -> str:
    return f"00000000-0000-{family}-8000-{index + 1:012d}"


def main() -> None:
    with transaction(platform=True) as connection:
        permission_ids = {}
        for index, (code, description) in enumerate(PERMISSION_DETAILS.items()):
            pid = stable_uuid("4001", index)
            connection.execute(
                text(
                    "INSERT INTO permissions(permission_id,permission_code,description) VALUES(:id,:code,:description) ON CONFLICT(permission_code) DO UPDATE SET description=excluded.description"
                ),
                {"id": pid, "code": code, "description": description},
            )
            permission_ids[code] = connection.execute(
                text("SELECT permission_id FROM permissions WHERE permission_code=:code"),
                {"code": code},
            ).scalar_one()
        for index, (code, (name, description)) in enumerate(ROLE_DETAILS.items()):
            rid = stable_uuid("4000", index)
            connection.execute(
                text(
                    "INSERT INTO roles(role_id,role_code,display_name,description,created_at) VALUES(:id,:code,:name,:description,now()) ON CONFLICT(role_code) DO UPDATE SET display_name=excluded.display_name,description=excluded.description"
                ),
                {"id": rid, "code": code, "name": name, "description": description},
            )
            actual = connection.execute(
                text("SELECT role_id FROM roles WHERE role_code=:code"), {"code": code}
            ).scalar_one()
            connection.execute(
                text("DELETE FROM role_permissions WHERE role_id=:id"), {"id": actual}
            )
            for permission in ROLE_GRANTS[code]:
                connection.execute(
                    text(
                        "INSERT INTO role_permissions(role_id,permission_id) VALUES(:role,:permission) ON CONFLICT DO NOTHING"
                    ),
                    {"role": actual, "permission": permission_ids[permission]},
                )
        for code, name in (
            ("CBSE", "Central Board of Secondary Education"),
            ("CISCE", "Council for the Indian School Certificate Examinations"),
        ):
            connection.execute(
                text(
                    "INSERT INTO boards(board_id,board_code,display_name,created_at) VALUES(:id,:code,:name,now()) ON CONFLICT(board_code) DO UPDATE SET display_name=excluded.display_name"
                ),
                {"id": str(uuid4()), "code": code, "name": name},
            )
        email = os.getenv("SEED_PLATFORM_ADMIN_EMAIL", "").strip().lower()
        password = os.getenv("SEED_PLATFORM_ADMIN_PASSWORD")
        if email or password:
            if not email or not password or len(password) < 14:
                raise ValueError(
                    "SEED_PLATFORM_ADMIN_EMAIL and a password of at least 14 characters must be supplied together"
                )
            user = connection.execute(
                text("SELECT user_id FROM users WHERE email=:email"), {"email": email}
            ).scalar()
            if not user:
                user = str(uuid4())
                connection.execute(
                    text(
                        "INSERT INTO users(user_id,email,password_hash,status,email_verified_at,created_at,updated_at) VALUES(:id,:email,:password,'ACTIVE',now(),now(),now())"
                    ),
                    {"id": user, "email": email, "password": hash_password(password)},
                )
                connection.execute(
                    text(
                        "INSERT INTO user_profiles(user_id,first_name,last_name,created_at,updated_at) VALUES(:id,:first,:last,now(),now())"
                    ),
                    {
                        "id": user,
                        "first": os.getenv("SEED_PLATFORM_ADMIN_FIRST_NAME", "Platform").strip()
                        or "Platform",
                        "last": os.getenv("SEED_PLATFORM_ADMIN_LAST_NAME", "Administrator").strip()
                        or "Administrator",
                    },
                )
            connection.execute(
                text(
                    "INSERT INTO platform_user_roles(user_id,role,created_at) VALUES(:id,'PLATFORM_SUPER_ADMIN',now()) ON CONFLICT DO NOTHING"
                ),
                {"id": user},
            )


if __name__ == "__main__":
    main()
